### PR TITLE
Add router API for invoice routing hints

### PR DESCRIPTION
**Pull request by t-bast**
The router can now return local channels, with all necessary information to build routing hints from these channels.

It's not eclair-core's responsibility to choose what channels to use in routing hints: this is really a wallet strategy, so each wallet should have its own heuristics to choose what channels to include (if any).

I think @btcontract was interested in these.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 1]
```diff
@@ -466,6 +493,11 @@ object Router {
   case object GetRoutingStateStreaming
   case object RoutingStateStreamingUpToDate
   case object GetRouterData
+  case object GetNodes
+  case object GetLocalChannels
+  case object GetChannels
+  case object GetChannelsMap
+  case object GetChannelUpdates
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:500]: You could have used the `replyTo` pattern for those new messages, but we may as well handle them all at once when the day comes.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 2]
```diff
@@ -162,19 +162,25 @@ class Router(val nodeParams: NodeParams, watcher: ActorRef, initialized: Option[
...
 
-    case Event(Symbol("nodes"), d) =>
+    case Event(GetNodes, d) =>
       sender ! d.nodes.values
       stay
 
-    case Event(Symbol("channels"), d) =>
+    case Event(GetLocalChannels, d) =>
+      val scids = d.graph.getIncomingEdgesOf(nodeParams.nodeId).map(_.desc.shortChannelId)
+      val localChannels = scids.flatMap(scid => d.channels.get(scid).map(c => LocalChannel(Right(c))).orElse(d.privateChannels.get(scid).map(c => LocalChannel(Left(c)))))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:171]: Slightly shorter version. I tried linebreaks, but it wasn't much more readable. We could save one iteration by removing the `.map(_.desc.shortChannelId)`.

```suggestion
      val localChannels = scids.flatMap(scid => d.channels.get(scid).map(Right(_)).orElse(d.privateChannels.get(scid).map(Left(_)).map(LocalChannel)))
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 3]
```diff
@@ -322,6 +328,27 @@ object Router {
       case Right(rcu) => updateChannelUpdateSameSideAs(rcu.channelUpdate)
     }
   }
+  case class LocalChannel(channel: Either[PrivateChannel, PublicChannel]) {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:331]: Should we have `PrivateChannel` and `PublicChannel` extend the same `trait`?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 4]
```diff
@@ -322,6 +328,27 @@ object Router {
...
+    def getRemoteUpdate(localNodeId: PublicKey): Option[ChannelUpdate] = channel.fold(
+      c => if (localNodeId == c.nodeId1) c.update_2_opt else c.update_1_opt,
+      c => if (localNodeId == c.ann.nodeId1) c.update_2_opt else c.update_1_opt
+    )
+
+    /** Create an invoice routing hint from that channel. Note that if the channel is private, the invoice will leak its existence. */
+    def toExtraHop(localNodeId: PublicKey): Option[ExtraHop] = getRemoteUpdate(localNodeId).map(remoteUpdate =>
+      ExtraHop(getRemoteNodeId(localNodeId), remoteUpdate.shortChannelId, remoteUpdate.feeBaseMsat, remoteUpdate.feeProportionalMillionths, remoteUpdate.cltvExpiryDelta)
+    )
+  }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:351]: It seems to me that we could make `remoteNodeId` part of the constructor. It is a bit strange that we need to provide more info for a class named `LocalChannel` to tell which side is the local one.

**Review by pm47 — COMMENTED** [review]
This doesn't "add a router API for invoice routing hints"?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:500]: Agreed, I'd rather change the whole router at once (or at least everything we can update in the router at once).

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 3 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:331]: Right, probably, I don't know why that wasn't done initially.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:351]: Agreed, I'll revisit that.

**Comment by t-bast** [comment]
> This doesn't "add a router API for invoice routing hints"?

Yes it does, but only to the router. It doesn't add it to the HTTP API as I think it's unnecessary, it's meant to be used by plugins for nodes that offer custodial wallet capabilities (for public nodes you usually don't need any routing hint, you have public channels).

**Comment by btcontract** [comment]
Yes, my rationale is that some nodes may start pruning smaller public channels at some point to keep graph manageable, so adding hints for smaller public channels can be useful.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Done in https://github.com/ACINQ/eclair/pull/1590/commits/273dd81593614f3323350c7ce0e3c52933841ca7, let me know what you think.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 3 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Done in https://github.com/ACINQ/eclair/pull/1590/commits/273dd81593614f3323350c7ce0e3c52933841ca7
The only downside is the methods that return a new instance of `PublicChannel/PrivateChannel` now return that shared abstraction, not sure whether it's an issue or not.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 4 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: I think it's better.

**Review by pm47 — APPROVED** [review]

**Pull request merged**